src/ess/nmx/mcstas_xml.py

```diff
@@ -0,0 +1,396 @@
+    ) -> 'SourceDesc':
+        """Create source description from xml component."""
+        source_xml = select_by_type_prefix(tree, 'sourceMantid-type')
+        location = select_by_tag(source_xml, 'location')
+
+        return cls(
+            component_type=source_xml.attrib['type'],
+            name=source_xml.attrib['name'],
+            position=_position_from_location(location, simulation_settings.length_unit),
+        )
+
+
+def _construct_pixel_ids(detector_descs: Tuple[DetectorDesc, ...]) -> sc.Variable:
+    """Pixel IDs for all detectors."""
+    intervals = [
+        (desc.id_start, desc.id_start + desc.total_pixels) for desc in detector_descs
+    ]
+    ids = [sc.arange('id', start, stop, unit=None) for start, stop in intervals]
+    return sc.concat(ids, 'id')
+
+
+def _pixel_positions(
+    detector: DetectorDesc, position_offset: sc.Variable
+) -> sc.Variable:
+    """Position of pixels of the ``detector``.
+
+    Position of each pixel is relative to the position_offset.
+    """
+    pixel_idx = sc.arange('id', detector.total_pixels)
+    n_row = sc.scalar(detector.num_fast_pixels_per_row)
+
+    pixel_n_slow = pixel_idx // n_row
+    pixel_n_fast = pixel_idx % n_row
+
+    fast_axis_steps = detector.fast_axis * detector.fast_step
+    slow_axis_steps = detector.slow_axis * detector.slow_step
+
+    return (
+        (pixel_n_slow * slow_axis_steps)
+        + (pixel_n_fast * fast_axis_steps)
+        + position_offset
+    )
+
+
+def _detector_pixel_positions(
+    detector_descs: Tuple[DetectorDesc, ...], sample: SampleDesc
+) -> sc.Variable:
+    """Position of pixels of all detectors."""
+
+    positions = [
+        _pixel_positions(detector, sample.position_from_sample(detector.position))
+        for detector in detector_descs
+    ]
+    return sc.concat(positions, 'panel')
+
+
+@dataclass
+class McStasInstrument:
+    simulation_settings: SimulationSettings
+    detectors: Tuple[DetectorDesc, ...]
+    source: SourceDesc
+    sample: SampleDesc
+
+    @classmethod
+    def from_xml(cls, tree: _XML) -> 'McStasInstrument':
+        """Create McStas instrument from xml."""
+        simulation_settings = SimulationSettings.from_xml(tree)
+
+        return cls(
+            simulation_settings=simulation_settings,
+            detectors=_collect_detector_descriptions(tree),
+            source=SourceDesc.from_xml(tree, simulation_settings),
+            sample=SampleDesc.from_xml(tree, simulation_settings),
+        )
+
+    def to_coords(self) -> dict[str, sc.Variable]:
+        """Extract coordinates from the McStas instrument description."""
+        slow_axes = [det.slow_axis for det in self.detectors]
+        fast_axes = [det.fast_axis for det in self.detectors]
+        origins = [
+            self.sample.position_from_sample(det.position) for det in self.detectors
+        ]
+        detector_dim = 'panel'
+
+        return {
+            'pixel_ids': _construct_pixel_ids(self.detectors),
```

Comment:
```suggestion
            'pixel_id': _construct_pixel_ids(self.detectors),
```
we generally go with singular names.